### PR TITLE
Testing improvements

### PR DESCRIPTION
**Pull request by jplatte**
I've been unhappy about our (integration) testing setup for a while, this should be a decent improvement while not requiring to rewrite large parts of the test suite at once.

### CODE REVIEW

File: testing/matrix-sdk-test/src/lib.rs [hunk 1]
```diff
@@ -1,18 +1,54 @@
...
+    api::{client::sync::sync_events::v3::Response as SyncResponse, IncomingResponse},
+    user_id, UserId,
+};
 use serde_json::Value as JsonValue;
 
+/// Create a `Raw<AnyTimelineEvent>` from arbitrary JSON.
+///
+/// Forwards all arguments to [`serde_json::json`].
+#[macro_export]
+macro_rules! timeline_event {
```
Comment (bnjbvr) [inline testing/matrix-sdk-test/src/lib.rs:14]: Small suggestion: could we name this `any_timeline_event`? I'm always confused about all the different timeline event types, so making it very explicit which it is would be helpful to me.

File: testing/matrix-sdk-test/src/lib.rs [hunk 2]
```diff
@@ -1,18 +1,54 @@
...
+            .unwrap()
+            .cast::<::ruma::events::AnyTimelineEvent>()
+    }
+}
+
+/// Create a `Raw<AnySyncTimelineEvent>` from arbitrary JSON.
+///
+/// Forwards all arguments to [`serde_json::json`].
+#[macro_export]
+macro_rules! sync_timeline_event {
```
Comment (bnjbvr) [inline testing/matrix-sdk-test/src/lib.rs:26]: same here?

File: testing/matrix-sdk-test/src/event_builder.rs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+        content: C,
+    ) -> Raw<AnySyncTimelineEvent> {
+        sync_timeline_event!({
+            "type": content.event_type(),
+            "content": content,
+            "event_id": event_id,
+            "sender": sender,
+            "origin_server_ts": self.next_server_ts(),
+        })
+    }
```
Comment (bnjbvr) [inline testing/matrix-sdk-test/src/event_builder.rs:99]: The existence and similarity of those two methods suggest a builder pattern, if you're so inclined (we could also revisit later when needs be).

File: crates/matrix-sdk-ui/src/timeline/tests/mod.rs [hunk 4]
```diff
@@ -165,31 +154,30 @@ impl TestTimeline {
...
     async fn handle_live_redaction(&self, sender: &UserId, redacts: &EventId) {
-        let ev = json!({
+        let ev = sync_timeline_event!({
             "type": "m.room.redaction",
             "content": {},
             "redacts": redacts,
             "event_id": EventId::new(server_name!("dummy.server")),
             "sender": sender,
-            "origin_server_ts": self.next_server_ts(),
+            "origin_server_ts": self.event_builder.next_server_ts(),
```
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/timeline/tests/mod.rs:173]: This is a bit of an abstraction leak here; does the `EventBuilder` provide a method to create such an event too, and could it have such a method if not?

File: crates/matrix-sdk-ui/src/timeline/tests/mod.rs [hunk 5]
```diff
@@ -200,10 +188,9 @@ impl TestTimeline {
             },
             "event_id": event_id,
             "sender": sender,
-            "origin_server_ts": self.next_server_ts(),
+            "origin_server_ts": self.event_builder.next_server_ts(),
```
Comment (bnjbvr) [inline crates/matrix-sdk-ui/src/timeline/tests/mod.rs:191]: ditto here

File: testing/matrix-sdk-test/src/event_builder.rs [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+    /// Set the next server timestamp.
+    ///
+    /// Timestamps will continue to increase by 1 (millisecond) from that value.
+    pub fn set_next_ts(&self, value: u64) {
+        self.next_ts.store(value, SeqCst);
+    }
+
+    pub fn make_message_event_with_id<C: MessageLikeEventContent>(
```
Comment (bnjbvr) [inline testing/matrix-sdk-test/src/event_builder.rs:60]: Overall design question: would it make sense to have one or two methods here that return builders that do follow a Builder pattern, so we don't have many small methods focused on a single thing, and can have more composability? (I would imagine the `EventBuilder` name would need to be replaced, as the type returned by those methods would be an `EventBuilder`)

**Review by bnjbvr — APPROVED** [review]
Looks good. Many small comments here and there, but this is test code so let's not spend too much time on it and add more features later as we see fit.

File: testing/matrix-sdk-test/src/lib.rs (hunk 1 above)
Comment (jplatte) [reply testing/matrix-sdk-test/src/lib.rs:14]: Ah but it returns a raw event, which can truly contain arbitrary events, not just the ones Ruma knows 😅

File: testing/matrix-sdk-test/src/event_builder.rs (hunk 3 above)
Comment (jplatte) [reply testing/matrix-sdk-test/src/event_builder.rs:99]: Ah but that requires separate finalization, which makes test code longer again :(

File: testing/matrix-sdk-test/src/event_builder.rs (hunk 6 above)
Comment (jplatte) [reply testing/matrix-sdk-test/src/event_builder.rs:60]: Postponing this to a follow-up PR.

**Pull request merged**